File: frontend/src/stores/reports.ts
import { defineStore } from "pinia";

import { endpoints } from "../api/endpoints";
import { apiClient } from "../api/http";

interface DailySales {
  date: string;
  total_orders: number;
  total_sales_cents: number;
  total_tax_cents: number;
  total_discount_cents: number;
}

interface OrderHistoryRow {
  order_id: number;
  order_no: string;
  status: string;
  order_type: string;
  table_id: number | null;
  opened_at: string;
  grand_total_cents: number | null;
}

interface SalesByDayRow {
  date: string;
  total_orders: number;
  total_sales_cents: number;
}

interface DateRangeParams {
  from_date?: string;
  to_date?: string;
}

export const useReportsStore = defineStore("reports", {
  state: () => ({
    daily: null as DailySales | null,
    history: [] as OrderHistoryRow[],
    salesByDay: [] as SalesByDayRow[],
    isLoading: false,
    error: "",
  }),
  actions: {
    async fetchDaily(date?: string) {
      this.isLoading = true;
      this.error = "";
      try {
        const { data } = await apiClient.get<DailySales>(`${endpoints.reports}/daily`, { params: { date } });
        this.daily = data;
        return data;
      } catch (_error) {
        this.error = "Failed to load daily report.";
        throw _error;
      } finally {
        this.isLoading = false;
      }
    },
    async fetchHistory(params?: DateRangeParams & { table_id?: number }) {
      this.isLoading = true;
      this.error = "";
      try {
        const queryParams = {
          from: params?.from_date,
          to: params?.to_date,
          table_id: params?.table_id,
        };
        const { data } = await apiClient.get<OrderHistoryRow[]>(`${endpoints.reports}/orders/history`, { params: queryParams });
        this.history = data;
        return data;
      } catch (_error) {
        this.error = "Failed to load order history.";
        throw _error;
      } finally {
        this.isLoading = false;
      }
    },
    async fetchSalesByDay(params?: DateRangeParams) {
      this.isLoading = true;
      this.error = "";
      try {
        const queryParams = {
          from: params?.from_date,
          to: params?.to_date,
        };
        const { data } = await apiClient.get<SalesByDayRow[]>(`${endpoints.reports}/sales-by-day`, { params: queryParams });
        this.salesByDay = data;
        return data;
      } catch (_error) {
        this.error = "Failed to load sales chart data.";
        throw _error;
      } finally {
        this.isLoading = false;
      }
    },
  },
});

File: backend/app/api/v1/routes/reports.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy import func
from sqlalchemy.orm import Session

from app.api.deps import get_db_session, require_roles
from app.core.roles import Role
from app.core.time import utcnow
from app.models.bill import Bill
from app.models.order import Order
from app.models.user import User
from app.schemas.report import DailySalesReport, OrderHistoryRow

router = APIRouter(prefix="/reports", tags=["reports"])


@router.get("/daily", response_model=DailySalesReport)
def daily_sales(
    date: str | None = Query(default=None),
    db: Session = Depends(get_db_session),
    _user: User = Depends(require_roles(Role.ADMIN, Role.CASHIER)),
) -> DailySalesReport:
    report_date = date or utcnow().date().isoformat()

    total_orders, total_sales, total_tax, total_discount = (
        db.query(
            func.count(Bill.id),
            func.coalesce(func.sum(Bill.grand_total_cents), 0),
            func.coalesce(func.sum(Bill.tax_cents), 0),
            func.coalesce(func.sum(Bill.discount_cents), 0),
        )
        .filter(func.date(Bill.issued_at) == report_date)
        .one()
    )

    return DailySalesReport(
        date=report_date,
        total_orders=int(total_orders or 0),
        total_sales_cents=int(total_sales or 0),
        total_tax_cents=int(total_tax or 0),
        total_discount_cents=int(total_discount or 0),
    )


@router.get("/orders/history", response_model=list[OrderHistoryRow])
def order_history(
    from_date: str | None = Query(default=None, alias="from"),
    to_date: str | None = Query(default=None, alias="to"),
    table_id: int | None = Query(default=None),
    db: Session = Depends(get_db_session),
    _user: User = Depends(require_roles(Role.ADMIN, Role.CASHIER, Role.WAITER)),
) -> list[OrderHistoryRow]:
    query = db.query(Order, Bill.grand_total_cents).outerjoin(Bill, Bill.order_id == Order.id)

    if from_date:
        query = query.filter(func.date(Order.opened_at) >= from_date)
    if to_date:
        query = query.filter(func.date(Order.opened_at) <= to_date)
    if table_id is not None:
        query = query.filter(Order.table_id == table_id)

    rows = query.order_by(Order.id.desc()).limit(500).all()
    result: list[OrderHistoryRow] = []
    for order, grand_total_cents in rows:
        result.append(
            OrderHistoryRow(
                order_id=order.id,
                order_no=order.order_no,
                status=order.status,
                order_type=order.order_type,
                table_id=order.table_id,
                opened_at=order.opened_at.isoformat(),
                grand_total_cents=grand_total_cents,
            )
        )
    return result
